### PR TITLE
920 project archive

### PR DESCRIPTION
**Pull request by martaswiatkowska**
Added new jira statuses: Approved, Archived, Closed.
Project can be archived.  

Fixed : #920

### CODE REVIEW

File: app/controllers/projects/archives_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class Projects::ArchivesController < ApplicationController
+  before_action :load_and_authorize_project!
+
+  def create
+    Project::Archive.new(@project).call
```
Comment (mkasztelnik) [inline app/controllers/projects/archives_controller.rb]: You should check if archiving was ended successfully and only then show "Project archived" notice.

File: app/models/project_item.rb [hunk 2]
```diff
@@ -9,7 +9,9 @@ class ProjectItem < ApplicationRecord
     in_progress: "in_progress",
     waiting_for_response: "waiting_for_response",
     ready: "ready",
-    rejected: "rejected"
+    rejected: "rejected",
+    closed: "closed",
```
Comment (mkasztelnik) [inline app/models/project_item.rb]: Two new states added here. Please make sure that these statuses are shown correctly on project item views (with correct name and color).

File: app/policies/project_policy.rb [hunk 3]
```diff
@@ -3,24 +3,29 @@
 class ProjectPolicy < ApplicationPolicy
   class Scope < Scope
     def resolve
-      scope.where(user: user)
+      scope.where(user: user, status: :active)
```
Comment (mkasztelnik) [inline app/policies/project_policy.rb]: It is correct? Maybe use should see 2 list - active and archived projects?

File: app/policies/project_policy.rb [hunk 4]
```diff
@@ -3,24 +3,29 @@
...
     end
   end
 
   def show?
-    owner?
+    owner? && record.active?
```
Comment (mkasztelnik) [inline app/policies/project_policy.rb]: Why the user cannot see archived projects?

File: app/services/project_item/ready.rb [hunk 5]
```diff
@@ -45,7 +47,7 @@ def ready_in_jira!
     end
 
     def update_status!
-      @project_item.new_status(status: :ready,
+      @project_item.new_status(status: :approved,
```
Comment (mkasztelnik) [inline app/services/project_item/ready.rb]: Please consult with @abacz how new states should be presented to the user.

File: app/views/projects/_project_layout.html.haml [hunk 6]
```diff
@@ -19,6 +19,9 @@
         - if policy(project).destroy?
           = link_to "Delete", project_path(project), method: :delete,
             data: { confirm: "Are you sure?" }, class: "btn btn-outline-danger btn-sm"
+        - if policy(project).archive?
+          = link_to "Archive", project_archive_path(project), method: :post, data: { confirm: "Are you sure?" },
```
Comment (mkasztelnik) [inline app/views/projects/_project_layout.html.haml]: Maybe more meaningful message explaining what archiving is doing and what are the consequences?

File: spec/features/projects_spec.rb [hunk 7]
```diff
@@ -9,11 +9,16 @@
...
   context "As logged in user" do
     let(:project_create) { double("Project::Create") }
+    let(:project_archive) { double("Project::Archive") }
     before { checkin_sign_in_as(user) }
     before(:each) {
       project_create_class_stub = class_double(Project::Create).
           as_stubbed_const(transfer_nested_constants: true)
       allow(project_create_class_stub).to receive(:new).and_return(project_create)
+
+      project_archive_class_stub = class_double(Project::Archive).
```
Comment (mkasztelnik) [inline spec/features/projects_spec.rb]: You can create a dedicated context for archiving and create this stubbing only in this scope.

File: spec/features/projects_spec.rb [hunk 8]
```diff
@@ -139,6 +144,51 @@
 
       expect(page).to have_text("You are not authorized to see this page")
     end
+
+    scenario "I cannot see archive button when are no services" do
```
Comment (mkasztelnik) [inline spec/features/projects_spec.rb]: `when the project has active orders`?

File: spec/features/projects_spec.rb [hunk 9]
```diff
@@ -139,6 +144,51 @@
...
+      project = create(:project, name: "First Project", user: user)
+
+      visit project_path(project)
+
+      expect(page).to_not have_selector(:link_or_button, "Archive")
+    end
+
+    scenario "I see archive button when all project_items are ended" do
```
Comment (mkasztelnik) [inline spec/features/projects_spec.rb]: What does it mean `ended`? It is a new order state, is its official name?

File: spec/services/jira/issue_updated_spec.rb [hunk 10]
```diff
@@ -16,25 +16,51 @@
...
-    described_class.new(project_item, changelog(to: jira_client.wf_done_id)).call
+    described_class.new(project_item, changelog(to: jira_client.wf_ready_id)).call
     last_change = project_item.statuses.last
 
     expect(last_change).to be_ready
     expect(last_change.message).to eq("Welcome!!!")
   end
 
+  it "uses service activate message when service become created" do
+    service = create(:service, activate_message: "Closed!!!")
```
Comment (mkasztelnik) [inline spec/services/jira/issue_updated_spec.rb]: Why activate message is used to show message when order is closed? (btw. what closed state means?)

File: spec/features/projects_spec.rb (hunk 9 above)
Comment (martaswiatkowska) [reply spec/features/projects_spec.rb]: Ended mean that service is rejected or closed. 
No its not official

File: app/views/projects/_project_layout.html.haml (hunk 6 above)
Comment (martaswiatkowska) [reply app/views/projects/_project_layout.html.haml]: @agpul? What to write here?

File: spec/services/jira/issue_updated_spec.rb (hunk 10 above)
Comment (martaswiatkowska) [reply spec/services/jira/issue_updated_spec.rb]: Its unnecessary

**Pull request merged**